lib/SPIRVProducerPass.cpp

```diff
@@ -4470,61 +4470,56 @@ void SPIRVProducerPass::GenerateInstruction(Instruction &I) {
       break;
     }
 
-    if (Callee->getName().startswith("spirv.atomic")) {
-      spv::Op opcode = StringSwitch<spv::Op>(Callee->getName())
-                           .Case("spirv.atomic_add", spv::OpAtomicIAdd)
-                           .Case("spirv.atomic_sub", spv::OpAtomicISub)
-                           .Case("spirv.atomic_exchange", spv::OpAtomicExchange)
-                           .Case("spirv.atomic_inc", spv::OpAtomicIIncrement)
-                           .Case("spirv.atomic_dec", spv::OpAtomicIDecrement)
-                           .Case("spirv.atomic_compare_exchange",
-                                 spv::OpAtomicCompareExchange)
-                           .Case("spirv.atomic_umin", spv::OpAtomicUMin)
-                           .Case("spirv.atomic_smin", spv::OpAtomicSMin)
-                           .Case("spirv.atomic_umax", spv::OpAtomicUMax)
-                           .Case("spirv.atomic_smax", spv::OpAtomicSMax)
-                           .Case("spirv.atomic_and", spv::OpAtomicAnd)
-                           .Case("spirv.atomic_or", spv::OpAtomicOr)
-                           .Case("spirv.atomic_xor", spv::OpAtomicXor)
-                           .Default(spv::OpNop);
-
-      //
-      // Generate OpAtomic*.
-      //
-      SPIRVOperandList Ops;
-
-      Ops << MkId(lookupType(I.getType()));
+    // Handle SPIR-V intrinsics
+    std::map<StringRef, spv::Op> SPVIntrToOpMap = {
```

Comment:
I'm concerned about the name overloading. Why didn't you leave unique names and change these to .StartsWith for the overloads? I'm worried someone could query an overload and get the wrong type. Do you have tests that use multiple variants of isnan for example?

Reply:
The `SPIRVProducerPass` was treating all of them in the same way which makes sense as all the required type information is conveyed by the IR types alone. FWIW, all the relevant OpenCL conformance tests pass with this change.

Reply:
Have you run a test that includes multiple variants of the same builtin? It feels like there would be a problem creating those function in LLVM.

Reply:
Oh, now I understand what you meant. It's probably broken. I'll look into it.

Reply:
Yeah, it's all horribly broken. https://github.com/google/clspv/pull/294 and https://github.com/google/clspv/pull/304 suffer from the same problem. I'll edit this PR to fix `isinf` and `isnan` and fix #294 as part of introducing a generic mechanism to deal with these. #304 probably shouldn't go in until then. Thanks for spotting this :).

Reply:
Actually, atomics have suffered from this issue since the first public commit so you may as well let #304 in. I'll fix all these.

Reply:
Thanks! I've merged #304

Reply:
Thanks! I've updated this PR. The updated code doesn't address any of the existing issues but should no longer add more of them.